From: Ben Skeggs <bskeggs@redhat.com>
Date: Thu, 19 Nov 2015 13:18:34 +1000
Subject: drm/nouveau/bios: return actual size of the buffer retrieved via _ROM
Origin: https://git.kernel.org/linus/954329412ea45ad6b509aa26f1441941fd432823
Bug: https://bugs.freedesktop.org/show_bug.cgi?id=90626
Bug-Debian: https://bugs.debian.org/772716

Fixes detection of a failed attempt at fetching the entire ROM image
in one-shot (a violation of the spec, that works a lot of the time).

Tested on a HP Zbook 15 G2.

Signed-off-by: Ben Skeggs <bskeggs@redhat.com>
---
 drivers/gpu/drm/nouveau/nouveau_acpi.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/gpu/drm/nouveau/nouveau_acpi.c
+++ b/drivers/gpu/drm/nouveau/nouveau_acpi.c
@@ -367,6 +367,7 @@ static int nouveau_rom_call(acpi_handle rom_handle, uint8_t *bios,
 		return -ENODEV;
 	}
 	obj = (union acpi_object *)buffer.pointer;
+	len = min(len, (int)obj->buffer.length);
 	memcpy(bios+offset, obj->buffer.pointer, len);
 	kfree(buffer.pointer);
 	return len;
